File: 29_DivideTwoIntegers.java
class Solution {
    public int divide(int dividend, int divisor) {
        if(dividend == Integer.MIN_VALUE && divisor == -1) return Integer.MAX_VALUE;
        if(dividend == Integer.MIN_VALUE && divisor == 1) return Integer.MIN_VALUE;

        boolean flag = (dividend > 0 && divisor > 0) || (dividend < 0 && divisor < 0);
        long res = 0;
        long num1 = Math.abs((long)dividend), num2 = Math.abs((long)divisor);

        while (num2 <= num1) {
            int shift = 0;
            while (num1 >= (num2 << shift)) {
                shift += 1;
            }
            num1 -= num2 << (shift - 1);
            res += 1 << (shift - 1);
        }

        return flag ? (int) res : (int) -res;
    }
}

File: c++/29_DivideTwoIntegers.cpp
class Solution {
public:
    int divide(int x, int y) {
        typedef long long LL;
        vector<LL> exp;
        bool flag = (x < 0 && y < 0) || (x > 0 && y > 0);

        LL a = abs(LL(x)), b = abs(LL(y));
        for (LL i = b; i <= a; i = i + i) exp.push_back(i);

        LL res = 0;
        for (int i = exp.size() - 1; i >= 0; i--) {
            if (a >= exp[i]) {
                a -= exp[i];
                res += 1ll << i;
            }
        }
        res = flag ? res : -res;
        if (res > INT_MAX || res < INT_MIN) res = INT_MAX;
        return res;
    }
};